### ISSUE TITLE
FlowId is not propagated properly

### ISSUE THREAD

**Issue by jonasp** [zalando-incubator/nakadi-java#193]
Setting the flow id as show in the example
```
EventMetadata em = new EventMetadata();
  .eid(UUID.randomUUID().toString())
  .occurredAt(OffsetDateTime.now())
  .flowId("decafbad");
```
does not result in the event metadata flow id to be equal to `decafbad`.

Nakadi expects the flow id in the header and not in the meta data.

**Comment by hgiddens** [comment]
I'll have a tilt at this.

**Comment by dehora** [comment]
> I'll have a tilt at this.

You're probably going to want to go through this interface [EventResource](https://github.com/zalando-incubator/nakadi-java/blob/master/nakadi-java-client/src/main/java/nakadi/EventResource.java), either mirroring the send methods to add `Map<String, String> headers`,  or add `EventResource flowId(String)` to the interface, and getting the data as far as [EventResourceReal.java#L123-L136](https://github.com/zalando-incubator/nakadi-java/blob/master/nakadi-java-client/src/main/java/nakadi/EventResourceReal.java#L123-L136)

**Comment by rcillo** [comment]
As a Nakadi maintainer, I'm very sorry about taking so long to solve this. Here is the PR in Nakadi that will change this behaviour making this unnecessary https://github.com/zalando/nakadi/pull/677

**Comment by dehora** [comment]
@rcillo are you happy this isn't just a client bug that should be fixed?

**Comment by rcillo** [comment]
@dehora my only fear is that someone relies on this behaviour. But that would be really, really surprising. Imagine that someone sets a value for "flow_id" in the body just to have it replaced by Nakadi?

So, yes, I think this should have been changed a long time ago.

**Comment by jonasp** [comment]
@rcillo what is the expected result if both header and flow id value are present because this is what nakadi-java-client will actually do (generated one in header).

**Comment by hgiddens** [comment]
One approach I'd considered was adding additional overloads e.g. `send(String, Collection<T>, String flowId)` to allow the flow ID for the request header to be specified explicitly.

Then in the original methods e.g. `send(String, Collection<T>)`, deriving the flow ID to use from the submitted events – if the events all have the same flow ID, use that, otherwise generate and use a new flow ID (i.e. the current behaviour). To do this I'd add a `FlowId` interface which `DataChangeEvent` and `BusinessEventMapped` (and potentially application-specific event implementations) implement to extract their contained flow IDs.

This potentially lets applications silently start doing the “right” thing, but is less straightforward than just adding an `EventResource flowId(String)` method – thoughts?

**Comment by rcillo** [comment]
@jonasp the flow-id present in the body will be preserved. The flow-id in the header will be ignored (we are just going to log it, of course). 

So it might be that a user have 2 different flow-id: one in the event body and another in the header. The header flow-id will be found in our logs and could be used for debugging. The event body flow id will be preserved as sent by the publisher.

**Comment by dehora** [comment]
#195 is available in 0.9.0
